SimG4CMS/PPS/src/PPSDiamondSD.cc

```diff
@@ -0,0 +1,304 @@
+// 2016 Nov
+///////////////////////////////////////////////////////////////////////////////
+#include "SimG4CMS/PPS/interface/PPSDiamondSD.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "SimG4Core/Notification/interface/TrackInformation.h"
+#include "SimG4Core/Geometry/interface/SensitiveDetectorCatalog.h"
+#include "SimG4Core/Physics/interface/G4ProcessTypeEnumerator.h"
+#include "SimG4Core/SensitiveDetector/interface/SensitiveTkDetector.h"
+#include "SimDataFormats/TrackingHit/interface/UpdatablePSimHit.h"
+#include "SimDataFormats/SimHitMaker/interface/TrackingSlaveSD.h"
+#include "G4SDManager.hh"
+#include "G4Step.hh"
+#include "G4Track.hh"
+#include "G4VProcess.hh"
+#include "G4ParticleDefinition.hh"
+#include "G4ParticleTypes.hh"
+#include "SimG4Core/Notification/interface/TrackInformation.h"
+#include "G4PhysicalConstants.hh"
+#include "G4SystemOfUnits.hh"
+#include <iostream>
+#include <vector>
+#include <string>
+
+PPSDiamondSD::PPSDiamondSD(const std::string& name,
+                           const edm::EventSetup& es,
+                           const SensitiveDetectorCatalog& clg,
+                           edm::ParameterSet const& p,
+                           const SimTrackManager* manager)
+    : SensitiveTkDetector(name, es, clg, p),
+      numberingScheme(nullptr),
+      hcID(-1),
+      theHC(nullptr),
+      currentHit(nullptr),
+      theTrack(nullptr),
+      currentPV(nullptr),
+      unitID(0),
+      preStepPoint(nullptr),
+      postStepPoint(nullptr),
+      eventno(0) {
+  collectionName.insert(name);
+  edm::ParameterSet m_Anal = p.getParameter<edm::ParameterSet>("PPSDiamondSD");
+  verbosity_ = m_Anal.getParameter<int>("Verbosity");
+
+  LogDebug("PPSSimDiamond") << "*******************************************************\n"
+                            << "*                                                     *\n"
+                            << "* Constructing a PPSDiamondSD  with name " << name << "\n"
+                            << "*                                                     *\n"
+                            << "*******************************************************"
+                            << "\n";
+
+  slave = new TrackingSlaveSD(name);
+
+  if (name == "CTPPSTimingHits") {
+    numberingScheme = dynamic_cast<PPSVDetectorOrganization*>(new PPSDiamondNumberingScheme());
+    edm::LogInfo("PPSSimDiamond") << "Find CTPPSDiamondHits as name";
+  } else {
+    edm::LogWarning("PPSSimDiamond") << "PPSDiamondSD: ReadoutName not supported\n";
+  }
+
+  edm::LogInfo("PPSSimDiamond") << "PPSDiamondSD: Instantiation completed";
+}
+
+PPSDiamondSD::~PPSDiamondSD() {
+  delete slave;
+  delete numberingScheme;
+}
+
+void PPSDiamondSD::Initialize(G4HCofThisEvent* HCE) {
+  LogDebug("PPSSimDiamond") << "PPSDiamondSD : Initialize called for " << name;
+
+  theHC = new PPSDiamondG4HitCollection(name, collectionName[0]);
```

Comment:
should be a unique_ptr constructed with make_unique

Reply:
As discussed above, G4Hit was kept as regular pointer. They are properly treated elsewere. All other occurence of pointers were move to unique_ptr.